File: Cpp/890_Find_and_Replace_Pattern.cpp
class Solution {
public:
    vector<string> findAndReplacePattern(vector<string>& words, string pattern) {
        vector<string> result;
        for(auto& word : words){
            if(match(word, pattern)) {
                result.push_back(word);
            }
        }
        
        return result;
    }
    
    bool match(string word, string pattern){
        unordered_map<char, char> m;
        unordered_set<char> s;
        for(int i=0; i<word.size(); ++i){
            if(m.count(word[i]) == 0){
                m[word[i]] = pattern[i];
                if(s.count(pattern[i])!=0) return false;
                s.insert(pattern[i]);
            } else if(m[word[i]] != pattern[i]){
                return false;
            }
        }
        
        return true;
    }
};

File: Kotlin/890_Find_and_Replace_Pattern.kt
class Solution {
    fun findAndReplacePattern(words: Array<String>, pattern: String): List<String> {
        val result = mutableListOf<String>()

        words.forEach{ word ->
            if(match(word, pattern)) {
                result.add(word)
            }
        }
        
        return result
    }
    
    private fun match(word: String, pattern: String): Boolean {
        val m = HashMap<Char, Char>()
        val s = HashSet<Char>()
        
        for(i in 0 until word.length) {
            if(!m.contains(word[i])){
                m[word[i]] = pattern[i]
                if(s.contains(pattern[i])) return false
                s.add(pattern[i])
            } else if(m[word[i]] != pattern[i]){
                return false
            }
        }
        
        return true
    }
}